// ominto-utils/src/schemas/api/accessApi.js
'use strict';

var Joi = require('joi');
const common = require('../common');

var emailRegister = Joi.object().keys({
  email: common.email.required(),
  isMobile: Joi.boolean().description('Optional flag which will return a much smaller activation code for ease of entering on touchscreen devices'),
  region: common.region,
  language: common.lang,
  currency: common.currency,
  referral: Joi.object().description('An object containing referral information'),
  policy_id: common.id
});

var emailActivate = Joi.object().keys({
  email: common.email.required(),
  code: Joi.string().max(255).required().description('Activation code generated during email register -- will be part of the activation link received via email'),
  password: common.password.required()
});

var emailLogin = Joi.object().keys({
  email: common.email.required(),
  password: common.password.required()
});

var resetPassword = Joi.object().keys({
  email: common.email.required()
});

var emailActivationResend = Joi.object().keys({
  email: common.email.required()
});

var phoneRegister = Joi.object().keys({
  phone: common.phone.required(),
  region: common.region,
  language: common.lang,
  currency: common.currency,
  referral: Joi.object().description('An object containing referral information'),
  policy_id: common.id
});

var phoneLogin = Joi.object().keys({
  phone: common.phone.required(),
  code: Joi.string().max(255).required(),
  region: common.region
});

var deviceRegister = Joi.object().keys({
  deviceId: common.deviceId.required(),
  region: common.region,
  language: common.lang,
  currency: common.currency,
  referral: Joi.object().description('An object containing referral information'),
  policy_id: common.id
});

var deviceConfirm = Joi.object().keys({
  deviceId: common.deviceId.required(),
  code: common.otpCode.required()
});

var deviceSetPin = Joi.object().keys({
  deviceId: common.deviceId.required(),
  code: common.otpCode.required(),
  oldPin: Joi.string().min(4).max(12),
  newPin: Joi.string().min(4).max(12).required()
});

var deviceLogin = Joi.object().keys({
  deviceId: common.deviceId.required(),
  code: common.otpCode.required(),
  pin: Joi.string().min(4).max(12).required()
});

var facebookDeviceLogin = Joi.object().keys({
  deviceId: common.deviceId.required(),
  code: common.otpCode.required(),
  facebook_id: Joi.string().required(),
  access_token: Joi.string().required()
});

var sso = Joi.object().keys({
  jwt: Joi.string().required(),
  redirect_url: Joi.string().allow('')
});

module.exports = {
  emailRegister: emailRegister,
  emailActivate: emailActivate,
  emailActivationResend: emailActivationResend,
  emailLogin: emailLogin,
  resetPassword: resetPassword,
  phoneRegister: phoneRegister,
  phoneLogin: phoneLogin,
  deviceRegister: deviceRegister,
  deviceConfirm: deviceConfirm,
  deviceSetPin: deviceSetPin,
  deviceLogin: deviceLogin,
  facebookDeviceLogin: facebookDeviceLogin,
  sso: sso
};
