// ominto-utils/src/schemas/api/cashbackApi.js
'use strict';

var Joi = require('joi');
const common = require('../common');

// OUTPUT VALIDATIONS

var _merchantAssets = Joi.object().keys({
  logo: common.asset.description("Asset object which holds information for the merchant's logo image"),
  cover: common.asset.description("Asset object which holds information for the merchant's cover image"),
  hero: common.asset.description("Asset object which holds information for the merchant's hero image")
});

var cashbackViewModel = Joi.object().keys({
  id: common.id,
  url: Joi.string().description('Ominto API URL which serves as the redirect link for this cashback'),
  siteUrl: Joi.string().description("The whitelabel-specific URL to view this cashback's website"),
  name: Joi.string().description('Display-friendly name of the cashback'),
  description: Joi.string().description('Brief description of the cashback'),
  merchantId: common.id,
  merchantName: Joi.string().description('Display-friendly name of the merchant who owns this cashback'),
  merchantAssets: _merchantAssets.description('An object holding all of the available image asset objects and their details'),
  minRate: Joi.number().min(0),
  maxRate: Joi.number().min(0),
  rate: Joi.string().description('The calculated percentage for either the authenticated user, or an unauthenticated user if no jwt is available -- please display this rate by default'),
  flat: Joi.string(),
  isDefault: Joi.boolean()
}).unknown(true);

var cashbackSimpleModel = Joi.object().keys({
  id: common.id,
  url: Joi.string().description('Ominto API URL which serves as the redirect link for this cashback'),
  name: Joi.string().description('Display-friendly name of the cashback'),
  description: Joi.string().description('Brief description of the cashback'),
  minRate: Joi.number().min(0),
  maxRate: Joi.number().min(0),
  rate: Joi.string().description('The calculated percentage for either the authenticated user, or an unauthenticated user if no jwt is available -- please display this rate by default'),
  flat: Joi.string(),
  isDefault: Joi.boolean()
}).unknown(true);

var cashbackViewModelOutput = Joi.object().keys({
  meta: Joi.object().unknown(true).description('A dynamic object containing metadata on the request/response'),
  cashbacks: Joi.array().items(cashbackViewModel).description('An array of cashback objects')
});

module.exports = {
  cashbackViewModel: cashbackViewModel,
  cashbackViewModelOutput: cashbackViewModelOutput,
  cashbackSimpleModel: cashbackSimpleModel
};
