// ominto-utils/src/schemas/api/categoryApi.js
'use strict';

var Joi = require('joi');
const common = require('../common');

// INPUT VALIDATIONS

var getCategoryById = Joi.object().keys({
  id: common.id.required()
});

var getCategoryBySlug = Joi.object().keys({
  slug: common.slug.required()
});

var getCategoryBySlugQuery = Joi.object().keys({
  language: common.lang.required(),
  region: common.region.required()
});

// OUTPUT VALIDATIONS

var _categoryAssets = Joi.object().keys({
  bannerWide: common.asset.description("Asset object which holds information for the category's banner image"),
  bannerEmail: common.asset.description("Asset object which holds information for the category's email banner image"),
  bannerMobile: common.asset.description("Asset object which holds information for the category's mobile banner image"),
  bannerEnvironment: common.asset.description("Asset object which holds information for the category's environment banner image"),
  collectionCard: common.asset.description("Asset object which holds information for the category's collection card image"),
  socialShare: common.asset.description("Asset object which holds information for the category's social share image"),
  twitterHeader: common.asset.description("Asset object which holds information for the category's Twitter header image"),
  facebookCover: common.asset.description("Asset object which holds information for the category's Facebook cover image")
});

var categoryViewModel = Joi.object().keys({
  id: common.id,
  name: Joi.string().description('Displayable name of the category'),
  slug: Joi.string().description('Resource slug intended to be used for a human-readable semantic URL for the category'),
  parentSlug: Joi.string(),
  siteUrl: Joi.string().description("The whitelabel-specific URL to view this category's website"),
  description: Joi.string().description('Brief description of the category'),
  details: Joi.string(),
  seoTitle: Joi.string(),
  seoKeywords: Joi.string(),
  displayMainTitle: Joi.string(),
  displaySubTitle: Joi.string(),
  facebookShareText: Joi.string(),
  twitterShareText: Joi.string(),
  assets: _categoryAssets.description('An object holding all of the available image asset objects and their details')
}).unknown(true);

var categorySimpleModel = Joi.object().keys({
  id: common.id,
  slug: Joi.string().description("Unique identifier for the category which are the 'slugified' category names"),
  name: Joi.string().description('English name of category used for display'),
  siteUrl: Joi.string().description("The whitelabel-specific URL to view this category's website")
}).unknown(true);

var categoryViewModelOutput = Joi.object().keys({
  meta: Joi.object().unknown(true).description('A dynamic object containing metadata on the request/response'),
  categories: Joi.array().items(categoryViewModel).description('An array of category objects')
});

module.exports = {
  getCategoryById: getCategoryById,
  getCategoryBySlug: getCategoryBySlug,
  getCategoryBySlugQuery: getCategoryBySlugQuery,
  categoryViewModelOutput: categoryViewModelOutput,
  categorySimpleModel: categorySimpleModel
};
