// ominto-utils/src/schemas/api/dubliNetworkApi.js
'use strict';

var Joi = require('joi');

// INCOMING VALIDATIONS

var pingback = {
  query: Joi.object().keys({
    code: Joi.string().required()
  }).unknown(true)
};

var setUserToPartnerProgramCoBrand = {
  type: 'json',
  body: Joi.object().keys({
    userId: Joi.number().integer().positive().required()
  })
};

var purchaseVouchers = {
  type: 'json',
  body: Joi.object().keys({
    amount: Joi.number().integer().positive().max(1000).required()
  })
};

var deleteVoucher = {
  params: Joi.object().keys({
    code: Joi.string().required()
  })
};

var getRedemptions = {
  query: Joi.object().keys({
    start: Joi.number().integer(), // unix timestamp
    end: Joi.number().integer()
  }).unknown(true)
};

var getTransactions = {
  query: Joi.object().keys({
    start: Joi.number().integer(),
    end: Joi.number().integer()
  }).unknown(true)
};

var getRegistrations = {
  query: Joi.object().keys({
    start: Joi.number().integer(),
    end: Joi.number().integer()
  }).unknown(true)
};

var getClicks = {
  query: Joi.object().keys({
    start: Joi.number().integer(),
    end: Joi.number().integer()
  }).unknown(true)
};

var getUserByEmail = {
  query: Joi.object().keys({
    email: Joi.string().required()
  }).unknown(true)
};


module.exports = {
  pingback: pingback,
  setUserToPartnerProgramCoBrand: setUserToPartnerProgramCoBrand,
  purchaseVouchers: purchaseVouchers,
  deleteVoucher: deleteVoucher,
  getRedemptions: getRedemptions,
  getTransactions: getTransactions,
  getRegistrations: getRegistrations,
  getClicks: getClicks,
  getUserByEmail: getUserByEmail
};
