// ominto-utils/src/schemas/api/regionApi.js
'use strict';

var Joi = require('joi');
const common = require('../common');

// INPUT VALIDATIONS

var getRegionById = Joi.object().keys({
  id: common.id.required()
});

var getRegionByIsoCode = Joi.object().keys({
  isocode: Joi.string().max(2).required().description('ISO 3166-1 alpha-2 country code')
});

// OUTPUT VALIDATIONS

var regionViewModel = Joi.object().keys({
  id: common.id,
  isocode: Joi.string().description('ISO 3166-1 alpha-2 country codes which serve as the unique identifier for regions'),
  name: Joi.string().description('English short country name officially used by ISO 3166')
}).unknown(true);

var regionViewModelOutput = Joi.object().keys({
  meta: Joi.object().unknown(true).description('A dynamic object containing metadata on the request/response'),
  regions: Joi.array().items(regionViewModel).description('An array of region objects')
});

module.exports = {
  getRegionById: getRegionById,
  getRegionByIsoCode: getRegionByIsoCode,
  regionViewModel: regionViewModel,
  regionViewModelOutput: regionViewModelOutput
};
