// ominto-utils/src/schemas/api/stripeApi.js
'use strict';

var Joi = require('joi');

var addCreditCard = Joi.object().keys({
  stripeToken: Joi.string().max(255).required(),
  isDefault: Joi.boolean()
});

var addBankAccount = Joi.object().keys({
  stripeToken: Joi.string().max(255).required(),
  isDefault: Joi.boolean()
});

var setDefaultCard = Joi.object().keys({
  source: Joi.string().max(255).required()
});

var setDefaultAcct = Joi.object().keys({
  source: Joi.string().max(255).required()
});

var startSubscription = Joi.object().keys({
  planId: Joi.string().max(255).required()
});

var oneTimeCharge = Joi.object().keys({
  source: Joi.string().max(255).required(),
  amount: Joi.number().integer().positive().max(9999999999).required().description('Currency amount in the lowest subdivision of the currency, e.g. cents for USD'),
  currency: Joi.string().lowercase().max(3).required().description('ISO 4217 three letter currency code'),
  descriptor: Joi.string()
});

var bankTransfer = Joi.object().keys({
  amount: Joi.number().integer().positive().max(9999999999).required().description('Currency amount in the lowest subdivision of the currency, e.g. cents for USD'),
  currency: Joi.string().lowercase().max(3).required().description('ISO 4217 three letter currency code')
});

module.exports = {
  addCreditCard: addCreditCard,
  addBankAccount: addBankAccount,
  setDefaultCard: setDefaultCard,
  setDefaultAcct: setDefaultAcct,
  startSubscription: startSubscription,
  oneTimeCharge: oneTimeCharge,
  bankTransfer: bankTransfer
};
