// ominto-utils/src/clients/communication.js
'use strict';

let config;

module.exports = function (configs) {
  config = configs;
  return {
    getUser: getUser,
    getMerchants: getMerchants,
    getStrings: getStrings
  };
};

function * getUser (userId, jwt, ctx) {
  var api = require('../utils/rest-client')({internalUrl: config.api.url, auth: {'bearer': jwt}}, false, ctx);
  var response = yield api.get('/comm/user/' + userId + '/', {}, ctx);
  ctx.assert(response.statusCode === 200, 500, 'Could not get user for communication', {res: response, userId: userId});
  return response.body;
}

function * getMerchants (userId, jwt, ctx) {
  var api = require('../utils/rest-client')({internalUrl: config.api.url, auth: {'bearer': jwt}}, false, ctx);
  var response = yield api.get('/comm/merchants/' + userId + '/', {}, ctx);
  ctx.assert(response.statusCode === 200, 500, 'Could not get merchants for communication', {res: response, userId: userId});
  return response.body;
}

function * getStrings (region, lang, scope, jwt, ctx) {
  var api = require('../utils/rest-client')({internalUrl: config.api.url, auth: {'bearer': jwt}}, false, ctx);
  var response = yield api.get('/strings/', {region: region, lang: lang, scope: scope}, ctx);
  ctx.assert(response.statusCode === 200, 500, 'Could not get strings for communication', {
    res: response,
    region: region,
    lang: lang,
    scope: scope
  });
  return response.body.strings;
}
