// ominto-utils/src/clients/mod.js
'use strict';

const _ = require('lodash');
let dataClient;

function * getModById (modId, ctx) {
  let result = yield dataClient.get('/searchModsForUser/' + modId + '/', null, ctx);
  ctx.assert(result.statusCode !== 404, 404, 'Mod not found');
  return result.body;
}

function * getModsForUser (userId, params, ctx) {
  let result = yield dataClient.get('/searchModsForUser/' + userId + '/', params, ctx);
  return result.body;
}

function * getVipMods (region, ctx) {
  let result = yield dataClient.get('/getVipMods/', {region: region}, ctx);
  return result.body;
}

function * getVipRedeemables (lang, region, ctx) {
  let result = yield dataClient.get('/getVipRedeemables/', {lang: lang, region: region}, ctx);
  return result.body;
}

function * deleteUserModifier (userId, modId, ctx) {
  let result = yield dataClient.del('/deleteModForUser/' + userId + '/' + modId + '/', null, ctx);
  return result.body;
}

function * reassignUserModifier (currentOwner, modId, targetOwner, ctx) {
  let result = yield dataClient.post('/reassignUserMod/' + currentOwner + '/' + modId + '/', {
    target: targetOwner
  }, ctx);
  return result.body;
}

function * updateUserModifier (user, mod, update, ctx) {
  let result = yield dataClient.post('/updateUserModifier/' + user + '/' + mod + '/', update, ctx);
  return result.body;
}

function * addRedeemableToUser (userId, redeemableId, ctx) {
  let result = yield dataClient.post('/addRedeemableToUser/', {userId: userId, redeemableId: redeemableId}, ctx);
  if (result.statusCode === 400) {
    ctx.throw(400, result.body.msg, {
      response: result
    });
  }
  return result.body;
}

function * searchRedeemedForUser (userId, params, ctx) {
  params = _.pick(params, ['date', 'region']);
  let result = yield dataClient.get('/searchRedeemedForUser/' + userId + '/', params, ctx);
  return result.body;
}

function * searchRedeemableForUser (userId, params, ctx) {
  params = _.pick(params, ['date', 'region']);
  let result = yield dataClient.get('/searchRedeemableForUser/' + userId + '/', params, ctx);
  return result.body;
}

module.exports = function init (config) {
  dataClient = require('../utils/rest-client')(config.data_api, false);
  return {
    getModById: getModById,
    getModsForUser: getModsForUser,
    getVipMods: getVipMods,
    getVipRedeemables: getVipRedeemables,
    deleteUserModifier: deleteUserModifier,
    reassignUserModifier: reassignUserModifier,
    updateUserModifier: updateUserModifier,
    addRedeemableToUser: addRedeemableToUser,
    searchRedeemedForUser: searchRedeemedForUser,
    searchRedeemableForUser: searchRedeemableForUser
  };
};
