// ominto-utils/src/utils/log.js
'use strict';
var firehoseLogger = require('./firehose-logger');
var bunyan = require('bunyan');
//var bunyanCW = require('./cloudwatch-logger');
var serializers = require('./custom-serializers');
var _ = require('lodash');

function createLogger (index, logStreamName, configs, options) {
  index = index.replace(/-/g, '_');
  var groupName = configs.env + '_' + index;
  //var region = configs.aws.default_region;
  options = _.defaultsDeep({
    _index: index,
    name: index,
    logGroupName: groupName,
    logStreamName: logStreamName.replace(/:/g, '-'),
    serializers: {
      req: bunyan.stdSerializers.req,
      res: bunyan.stdSerializers.res,
      err: serializers.err
    }
  }, options || {}, {
    streams: []
  });

  var logger;

  if (configs && (configs.env === 'stage' || configs.env === 'prod')) {
    firehoseLogger.initialize(configs);
    logger = firehoseLogger.createLogger(options, configs.env);
  } else {
    options.streams.push({
      stream: process.stdout,
      level: bunyan.TRACE
    });
    logger = bunyan.createLogger(options);
  }

  return logger;
}

module.exports = {
  createLogger: createLogger
};
