// ominto-utils/src/utils.js
'use strict';

module.exports = {
  addSubIdToTrackingUrl: require('./utils/add-subid-to-tracking-url'),
  appendReqId: require('./utils/req-and-session').getReqId,
  checkApiResponse: require('./utils/check-api-response'),
  clients: require('./clients/clients'),
  createEnvelope: require('./utils/create-envelope').createEnvelope,
  createEnvelopeFromLogger: require('./utils/create-envelope').fromLogger,
  createTrigger: require('./utils/create-envelope').createTrigger,
  createTriggerFromLogger: require('./utils/create-envelope').createTriggerFromLogger,
  currency: require('./utils/currency'),
  date: require('./utils/date'),
  duck: require('./utils/duck-checks'),
  findAvailableSlug: require('./utils/find-available-slug'),
  firehoseLogger: require('./utils/firehose-logger'),
  generateCode: require('./utils/generate-activation-code'),
  getDataClient: require('./utils/get-data-client'),
  getIntegerAmountForCurrency: require('./utils/currency').getIntegerAmountForCurrency,
  getMinorAmountForTransactionCurrency: require('./utils/currency').getMinorAmountForTransactionCurrency,
  getSessionId: require('./utils/req-and-session').getSessionId,
  handleError: require('./utils/handle-api-error'),
  hash: require('./utils/hash'),
  healthcheck: require('./utils/healthcheck'),
  lambdaHandler: require('./utils/lambda-handler'),
  log: require('./utils/log'),
  metrics: require('./utils/metrics-client'),
  mustache: require('./utils/mustache-simple'),
  oauth: require('./utils/oauth'),
  promisePiggyback: require('./utils/promise-piggyback'),
  promiseRateLimiter: require('./utils/promise-rate-limiter'),
  query: require('./utils/query-builder'),
  reqAndSession: require('./utils/req-and-session'),
  restClient: require('./utils/rest-client'),
  sanitize: require('./utils/sanitize'),
  schemas: require('./schemas'),
  search: require('./utils/elastic-search-client'),
  serializers: require('./utils/custom-serializers'),
  setIntoResponseBody: require('./utils/set-into-response-body'),
  splitObjectBySchemaDef: require('./utils/split-object-by-schema-def'),
  stripTags: require('./utils/strip-tags'),
  subId: require('./utils/sub-id'),
  url: require('./utils/url'),
  user: {
    modifiers: require('./utils/user/user-modifier'),
    points: require('./utils/user/points'),
    utm: require('./utils/user/utm')
  },
  viewModelMagic: require('./utils/mustache-simple').viewModelMagic,
  whitelabelInfoByName: require('./utils/whitelabel-info').whitelabelInfoByName,
  whitelabelInfoByDomain: require('./utils/whitelabel-info').whitelabelInfoByDomain,
  masspayStatus: require('./utils/masspay-status'),

  // data client calls
  userClient: require('./clients/user'),
  deviceClient: require('./clients/device'),
  kinesisClient: require('./clients/kinesisEvent'),
  paymentClient: require('./clients/payment'),
  merchantClient: require('./clients/merchant')
};
