// ominto-utils/src/schemas/common.js
'use strict';

const Joi = require('joi');

const id = Joi.number().integer().min(0).description('Auto-incrementing integer ID for internal use only');
const whitelabel = Joi.string().valid('dubli', 'infinia', 'ominto', 'inficashback','bsprewards','partner', 'democashback' ,'cashbackarabia');
const name = Joi.string().max(255);
const slug = Joi.string().max(255).regex(/^[a-z0-9-]+$/, 'Slug regex: only lowercase alphanumeric and hyphen characters').description('Resource slug intended to be used for a human-readable semantic URL');
const description = Joi.string().allow(null, '');
const active = Joi.number().integer().max(1).description('Active flag which is represented as a tinyint -- only possible values should be 0 or 1');
const regionChanged = Joi.number().integer().max(1).description('RegionChanged flag which is represented as a tinyint -- only possible values should be 0 or 1');
const newsLetterFollowed = Joi.number().integer().max(1).description('newsLetterFollowed flag which is represented as a tinyint -- only possible values should be 0 or 1');
const region = Joi.string().lowercase().max(2).description('ISO 3166-1 alpha-2 country code');
const lang = Joi.string().lowercase().max(2).description('ISO 639-1 two letter language code');
const currency = Joi.string().lowercase().max(3).description('ISO 4217 three letter currency code');
const email = Joi.string().email().max(255).description('RFC 5321 compatible');
const password = Joi.string().min(8).max(100);
const phone = Joi.string().max(16).description("E.164 international phone number format, e.g. '+15554449876' for US -- [+][country code][subscriber number including area code]");
const deviceId = Joi.string().max(100).description('Unique identifier for a device which is generated by the client');
const otpCode = Joi.string().max(6).description('OTP code generated from the secret key which was returned during device register -- will be validated against secret key on server side');
const gateway = Joi.string().valid('stripe', 'paypal', 'dwolla', 'payu','novalnet', 'dohabank', 'coinbase', 'infiplanet', 'ominto', 'inreft', 'paytm', 'epaisa', 'creditcard', 'bankaccount', 'ewallet', 'payoneer');
const createdVia = Joi.string().valid('importer', 'admin');

const asset = Joi.object().keys({
  source: Joi.string().description('URL to the location of the asset image source'),
  original_source: Joi.string().description('URL to the location of the original asset image source'),
  small: Joi.string().description('URL to the location of the small-sized asset image'),
  medium: Joi.string().description('URL to the location of the medium-sized asset image'),
  large: Joi.string().description('URL to the location of the large-sized asset image'),
  dpi_1: Joi.string(),
  dpi_1325: Joi.string(),
  dpi_15: Joi.string(),
  dpi_18: Joi.string(),
  dpi_2: Joi.string(),
  dpi_24: Joi.string(),
  dpi_26: Joi.string(),
  dpi_3: Joi.string(),
  dpi_35: Joi.string(),
  dpi_4: Joi.string(),
  bgColor: Joi.string().description("Hex code of the asset's background color"),
  primaryColor: Joi.string().description("Hex code of the asset's primary color"),
  last_uploaded: Joi.date(),
  last_processed: Joi.date()
});

module.exports = {
  id: id,
  whitelabel: whitelabel,
  name: name,
  slug: slug,
  description: description,
  active: active,
  regionChanged: regionChanged,
  region: region,
  newsLetterFollowed:newsLetterFollowed,
  lang: lang,
  currency: currency,
  email: email,
  password: password,
  phone: phone,
  deviceId: deviceId,
  otpCode: otpCode,
  gateway: gateway,
  asset: asset,
  createdVia: createdVia
};
