// ominto-utils/src/schemas/index.js
'use strict';
module.exports = {
  event: require('./event'),

  // DATA schemas
  admin_user: require('./data/admin_user'),
  admin_user_authorization: require('./data/admin_user_authorization'),
  asset: require('./data/asset'),
  category: require('./data/category'),
  currency: require('./data/currency'),
  device: require('./data/device'),
  faq: require('./data/faq'),
  faqEntry: require('./data/faq-entry'),
  merchant: require('./data/merchant'),
  merchant_cashback: require('./data/merchant_cashback'),
  merchant_deal: require('./data/merchant_deal'),
  merchant_product: require('./data/merchant_product'),
  mods: require('./data/mods'),
  novalnet: require('./data/novalnet'),
  outclick: require('./data/outclick'),
  outclick_feedback: require('./data/outclick_feedback'),
  page: require('./data/page'),
  payment: require('./data/payment'),
  payment_method: require('./data/payment_method'),
  pending_withdrawals: require('./data/pending_withdrawals'),
  points: require('./data/points'),
  policies: require('./data/policies'),
  region: require('./data/region'),
  segment: require('./data/segment'),
  selector: require('./data/selector'),
  share_links: require('./data/share_links'),
  shopping_event: require('./data/shopping_event'),
  strings: require('./data/strings'),
  transactions: require('./data/transactions'),
  user_acceptances: require('./data/user_acceptances'),
  user_points: require('./data/user_points'),
  users: require('./data/users'),
  users_referral: require('./data/users_referral'),
  ba_landingPage: require('./data/ba_landingPage'),
  voucher: require('./data/voucher'),

  // LOG SCHEMAS FOR DATA
  merchantLog: require('./data/log_tables/merchantLog'),
  merchantDealLog: require('./data/log_tables/merchantDealLog'),
  merchantCashBackLog: require('./data/log_tables/merchantCashBackLog'),
  categoryLog: require('./data/log_tables/categoryLog'),

  // API schemas
  accessApi: require('./api/accessApi'),
  cashbackApi: require('./api/cashbackApi'),
  categoryApi: require('./api/categoryApi'),
  dealApi: require('./api/dealApi'),
  dubliNetworkApi: require('./api/dubliNetworkApi'),
  dwollaApi: require('./api/dwollaApi'),
  merchantApi: require('./api/merchantApi'),
  paymentApi: require('./api/paymentApi'),
  policyApi: require('./api/policyApi'),
  regionApi: require('./api/regionApi'),
  searchApi: require('./api/searchApi'),
  shoppingEventApi: require('./api/shoppingEventApi'),
  stripeApi: require('./api/stripeApi'),
  trackerApi: require('./api/trackerApi'),
  transactionsApi: require('./api/transactionsApi'),
  userApi: require('./api/userApi'),
  webhooksApi: require('./api/webhooksApi')
};
