// ominto-utils/src/utils/user/utm.js
'use strict';

const _ = require('lodash');

function extractUtmParamsFromUser (user) {
  return {
    source: _.get(user, 'referral.source'),
    medium: _.get(user, 'referral.medium'),
    term: _.get(user, 'referral.term'),
    content: _.get(user, 'referral.content'),
    name: _.get(user, 'referral.name'),
    baId: _.get(user, 'referral.baid') || _.get(user, 'referral.baId'),
    Refer_BusinessAssignedBA: _.get(user, 'Refer.BusinessAssignedBA') || _.get(user, 'Refer.BusinessAssignedba'),
    baId2: _.get(user, 'referral.baid2') || _.get(user, 'referral.baId2'),
    Refer_BusinessAssignedBA2: _.get(user, 'Refer.BusinessAssignedBA2') || _.get(user, 'Refer.BusinessAssignedba2'),
    baId3: _.get(user, 'referral.baid3') || _.get(user, 'referral.baId3'),
    Refer_BusinessAssignedBA3: _.get(user, 'Refer.BusinessAssignedBA3') || _.get(user, 'Refer.BusinessAssignedba3'),
    baId4: _.get(user, 'referral.baid4') || _.get(user, 'referral.baId4'),
    Refer_BusinessAssignedBA4: _.get(user, 'Refer.BusinessAssignedBA4') || _.get(user, 'Refer.BusinessAssignedba4'),
    baId5: _.get(user, 'referral.baid5') || _.get(user, 'referral.baId5'),
    Refer_BusinessAssignedBA5: _.get(user, 'Refer.BusinessAssignedBA5') || _.get(user, 'Refer.BusinessAssignedba5'),
    baId6: _.get(user, 'referral.baid6') || _.get(user, 'referral.baId6'),
    Refer_BusinessAssignedBA6: _.get(user, 'Refer.BusinessAssignedBA6') || _.get(user, 'Refer.BusinessAssignedba6'),
    baId7: _.get(user, 'referral.baid7') || _.get(user, 'referral.baId7'),
    Refer_BusinessAssignedBA7: _.get(user, 'Refer.BusinessAssignedBA7') || _.get(user, 'Refer.BusinessAssignedba7'),
    baId8: _.get(user, 'referral.baid8') || _.get(user, 'referral.baId8'),
    Refer_BusinessAssignedBA8: _.get(user, 'Refer.BusinessAssignedBA8') || _.get(user, 'Refer.BusinessAssignedba8'),
    subId1: _.get(user, 'referral.subid1') || _.get(user, 'referral.subId1'),
    subId2: _.get(user, 'referral.subid2') || _.get(user, 'referral.subId2'),
    subId3: _.get(user, 'referral.subid3') || _.get(user, 'referral.subId3'),
    subId4: _.get(user, 'referral.subid4') || _.get(user, 'referral.subId4'),
    linkId: _.get(user, 'referral.linkid') || _.get(user, 'referral.linkId')
  };
}

module.exports = extractUtmParamsFromUser;
